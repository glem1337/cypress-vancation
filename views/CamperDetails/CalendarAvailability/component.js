import React from 'react';
import { FormattedMessage } from 'react-intl';
import ReactCalendar from 'react-calendar';
import classnames from 'classnames';
import { Spin } from 'antd';
import { LoadingOutlined } from '@ant-design/icons';

import isMobileView from 'utils/breakpoints/isMobileView';
import useCalendarAvailability from 'utils/hooks/useCalendarAvailability';

import useContainer from './hook';

const CalendarAvailability = () => {
  const {
    containerRef,
    searchParams,
  } = useContainer();

  const {
    formatShortWeekday,
    tileContent,
    tileDisabled,
    onDateRangeChange,
    isCalendarFetching,
    onActiveStartDateChange,
  } = useCalendarAvailability();

  return (
    <div
      id="details-availability"
      ref={containerRef}
    >
      <p className="text-headline mb-24">
        <FormattedMessage id="shared.availability" />
      </p>
      <div
        id="campervan-availability__calendar-wrapper"
        className={
          classnames(
            'campervan-availability__calendar-wrapper',
            {
              'campervan-availability__calendar-wrapper--loading': isCalendarFetching,
            },
          )
        }
      >
        <ReactCalendar
          onChange={onDateRangeChange}
          onActiveStartDateChange={onActiveStartDateChange}
          selectRange
          value={searchParams?.dateRange}
          showDoubleView={!isMobileView()}
          formatShortWeekday={formatShortWeekday}
          tileContent={tileContent}
          tileDisabled={tileDisabled}
          calendarType="US"
        />
        {isCalendarFetching && (
          <div className="campervan-availability__calendar-loader-wrapper">
            <Spin
              indicator={<LoadingOutlined style={{ fontSize: 88 }} spin />}
              className="search-page__spin"
            />
          </div>
        )}
      </div>
    </div>
  );
};

export default CalendarAvailability;
