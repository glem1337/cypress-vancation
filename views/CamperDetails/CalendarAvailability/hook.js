import { useRef, useEffect } from 'react';
import { useRouter } from 'next/router';
import * as R from 'ramda';
import { useDispatch, useSelector } from 'react-redux';

import { fetchCamperCalendar } from 'state/concepts/camper/actions';
import { searchDestinationParamsSelector } from 'state/concepts/search-destinations/selectors';

function useContainer() {
  const dispatch = useDispatch();

  const containerRef = useRef();

  const observer = useRef();

  const isDataRequested = useRef(false);

  const router = useRouter();

  const searchParams = useSelector(searchDestinationParamsSelector);

  const camperId = R.path(['query', 'camper_id'], router);

  /**
   * Handle intersect
   */
  const handleIntersect = (entries) => {
    const isIntersecting = R.path(['0', 'isIntersecting'], entries);

    if (!isIntersecting || isDataRequested.current) {
      return false;
    }

    dispatch(fetchCamperCalendar({ camperId }));

    return true;
  };

  /**
   * Unmount
   */
  const unmount = () => {
    if (!observer?.current?.unobserve || !containerRef?.current) {
      return false;
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
    observer.current.unobserve(containerRef.current);

    return true;
  };

  /**
   * Mounting.
   */
  useEffect(() => {
    if (!containerRef?.current) {
      return null;
    }

    observer.current = new IntersectionObserver(handleIntersect, {
      rootMargin: '700px',
    });
    observer.current.observe(containerRef.current);

    return unmount;
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return {
    containerRef,
    searchParams,
    handleIntersect,
    unmount,
  };
}

export default useContainer;
