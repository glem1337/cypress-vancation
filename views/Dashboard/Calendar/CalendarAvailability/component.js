import React from 'react';
import { Button, Popover } from 'antd';
import { FormattedMessage } from 'react-intl';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { AVAILABILITY_MODE } from 'constants/calendar';
import blockDatesPopoverContent from 'utils/calendar/blockDatesPopoverContent';
import unBlockDatesPopoverContent from 'utils/calendar/unBlockDatesPopoverContent';

import MinNightStay from './MinNightStay';
import BaseDiscounts from './BaseDiscounts';
import BaseDiscountsSingle from './BaseDiscountsSingle';
import NightlyRate from './NightlyRate';
import NightlyRateSingle from './NightlyRateSingle';

const CalendarAvailabilityComponent = ({
  availabilityMode,
  camperId,
  isBlockedPopoverVisible,
  isUnBlockedPopoverVisible,
  blockReject,
  blockConfirm,
  blockPrepare,
  isBlockedLoading,
  isUnBlockedLoading,
  unBlockReject,
  unBlockConfirm,
  unBlockPrepare,
}) => (
  <>
    <div className="btns-group mb-8">
      <Popover
        visible={isUnBlockedPopoverVisible}
        content={unBlockDatesPopoverContent({
          unBlockReject,
          unBlockConfirm,
        })}
        trigger="click"
        placement="bottom"
      >
        <Button
          className={classNames('dashboard-calendar__availability-button', {
            'btns-group-active': availabilityMode === AVAILABILITY_MODE.AVAILABLE,
          })}
          size="small"
          onClick={unBlockPrepare}
          loading={isUnBlockedLoading}
        >
          <FormattedMessage id="shared.available" />
        </Button>
      </Popover>
      <Popover
        visible={isBlockedPopoverVisible}
        content={blockDatesPopoverContent({
          blockReject,
          blockConfirm,
        })}
        trigger="click"
        placement="bottom"
      >
        <Button
          className={classNames('dashboard-calendar__availability-button', {
            'btns-group-active': availabilityMode === AVAILABILITY_MODE.BLOCKED,
          })}
          size="small"
          onClick={blockPrepare}
          loading={isBlockedLoading}
        >
          <FormattedMessage id="shared.blocked" />
        </Button>
      </Popover>
    </div>
    {(
      availabilityMode === AVAILABILITY_MODE.AVAILABLE
      || availabilityMode === AVAILABILITY_MODE.BLOCKED
    ) && (
      <>
        <NightlyRate camperId={camperId} />
        <NightlyRateSingle camperId={camperId} />
        <BaseDiscounts camperId={camperId} />
        <BaseDiscountsSingle camperId={camperId} />
        <MinNightStay camperId={camperId} />
      </>
    )}
  </>
);

CalendarAvailabilityComponent.propTypes = {
  availabilityMode: PropTypes.string.isRequired,
  camperId: PropTypes.string.isRequired,
  isBlockedPopoverVisible: PropTypes.bool.isRequired,
  isUnBlockedPopoverVisible: PropTypes.bool.isRequired,
  blockReject: PropTypes.func.isRequired,
  blockConfirm: PropTypes.func.isRequired,
  blockPrepare: PropTypes.func.isRequired,
  unBlockReject: PropTypes.func.isRequired,
  unBlockConfirm: PropTypes.func.isRequired,
  unBlockPrepare: PropTypes.func.isRequired,
  isBlockedLoading: PropTypes.bool.isRequired,
  isUnBlockedLoading: PropTypes.bool.isRequired,
};

export default CalendarAvailabilityComponent;
