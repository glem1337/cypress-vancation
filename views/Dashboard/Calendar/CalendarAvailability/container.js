import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as R from 'ramda';

import { AVAILABILITY_MODE } from 'constants/calendar';
import { setAvailabilityMode as setAvailabilityModeAction } from 'state/concepts/calendar/actions';
import { availabilityModeSelector, selectedSlotsSelector } from 'state/concepts/calendar/selectors';
import { createCalendarBlockedPeriodsEndpoint, deleteCalendarBlockedPeriodsEndpoint } from 'state/concepts/camper/endpoints';
import {
  createBlockedPeriods as createBlockedPeriodsAction,
  deleteBlockedPeriods as deleteBlockedPeriodsAction,
} from 'state/concepts/camper/actions';
import { loadingSelector } from 'state/data/selectors';

import CalendarAvailabilityComponent from './component';

class CalendarAvailability extends React.Component {
  static propTypes = {
    setAvailabilityMode: PropTypes.func.isRequired,
    availabilityMode: PropTypes.string.isRequired,
    createBlockedPeriods: PropTypes.func.isRequired,
    deleteBlockedPeriods: PropTypes.func.isRequired,
    camperId: PropTypes.string.isRequired,
    selectedSlots: PropTypes.shape().isRequired,
    isBlockedLoading: PropTypes.bool,
    isUnBlockedLoading: PropTypes.bool,
  }

  static defaultProps = {
    isBlockedLoading: false,
    isUnBlockedLoading: false,
  }

  state = {
    isBlockedPopoverVisible: false,
    isUnBlockedPopoverVisible: false,
  }

  changeAvailabilityMode = (mode) => () => {
    const { setAvailabilityMode } = this.props;

    setAvailabilityMode(mode);
  }

  blockReject = () => {
    this.setState({ isBlockedPopoverVisible: false });
  }

  blockConfirm = () => {
    const { setAvailabilityMode, createBlockedPeriods, camperId, selectedSlots } = this.props;

    this.setState({ isBlockedPopoverVisible: false });
    setAvailabilityMode(AVAILABILITY_MODE.BLOCKED);

    createBlockedPeriods({
      camperId,
      startDate: R.head(selectedSlots.slots),
      endDate: R.last(selectedSlots.slots),
    });
  }

  blockPrepare = () => {
    const { isUnBlockedPopoverVisible, isBlockedPopoverVisible } = this.state;
    const { isBlockedLoading, isUnBlockedLoading } = this.props;

    if (
      isUnBlockedPopoverVisible
      || isBlockedPopoverVisible
      || isBlockedLoading
      || isUnBlockedLoading
    ) {
      return;
    }

    this.setState({ isBlockedPopoverVisible: true });
  }

  unBlockReject = () => {
    this.setState({ isUnBlockedPopoverVisible: false });
  }

  unBlockConfirm = () => {
    const { setAvailabilityMode, deleteBlockedPeriods, camperId, selectedSlots } = this.props;

    this.setState({ isUnBlockedPopoverVisible: false });
    setAvailabilityMode(AVAILABILITY_MODE.AVAILABLE);

    deleteBlockedPeriods({
      camperId,
      startDate: R.head(selectedSlots.slots),
      endDate: R.last(selectedSlots.slots),
    });
  }

  unBlockPrepare = () => {
    const { isUnBlockedPopoverVisible, isBlockedPopoverVisible } = this.state;
    const { isBlockedLoading, isUnBlockedLoading } = this.props;

    if (
      isUnBlockedPopoverVisible
      || isBlockedPopoverVisible
      || isBlockedLoading
      || isUnBlockedLoading
    ) {
      return;
    }

    this.setState({ isUnBlockedPopoverVisible: true });
  }

  render() {
    return (
      <CalendarAvailabilityComponent
        {...this.props}
        {...this.state}
        changeAvailabilityMode={this.changeAvailabilityMode}
        blockReject={this.blockReject}
        blockConfirm={this.blockConfirm}
        blockPrepare={this.blockPrepare}
        unBlockReject={this.unBlockReject}
        unBlockConfirm={this.unBlockConfirm}
        unBlockPrepare={this.unBlockPrepare}
      />
    );
  }
}

const mapStateToProps = (state) => ({
  availabilityMode: availabilityModeSelector(state),
  selectedSlots: selectedSlotsSelector(state),
  isBlockedLoading: loadingSelector(state, createCalendarBlockedPeriodsEndpoint.endpoint),
  isUnBlockedLoading: loadingSelector(state, deleteCalendarBlockedPeriodsEndpoint.endpoint),
});

const mapDispatchToProps = {
  setAvailabilityMode: setAvailabilityModeAction,
  createBlockedPeriods: createBlockedPeriodsAction,
  deleteBlockedPeriods: deleteBlockedPeriodsAction,
};

export { CalendarAvailability as CalendarAvailabilityContainer };
export default connect(mapStateToProps, mapDispatchToProps)(CalendarAvailability);
