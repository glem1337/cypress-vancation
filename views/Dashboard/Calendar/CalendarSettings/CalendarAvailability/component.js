import classNames from 'classnames';
import { Button, Form } from 'antd';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { Field } from 'formik';

import { CALENDAR_AVAILABILITY_OPTIONS } from 'constants/calendar';
import SelectField from 'views/AddNewCamper/CamperDetails/SelectField';

import Footer from '../../Footer';
import BackButton from '../../BackButton';

const CalendarAvailability = (props) => {
  const {
    toggleOpenedState,
    isOpened,
    isBackPopoverVisible,
    isDiscardPopoverVisible,
    isLoading,
    isValid,
    submitForm,
    isSavePopoverVisible,
    showPopoverOrBack,
    closePopover,
    showPopover,
    closePopoversAndBack,
    closePopoversAndSave,
  } = props;

  return (
    <div className="calendar-listing__side-main-item">
      <Button className="ant-btn-slideout" onClick={toggleOpenedState}>
        <i className="icon icon-calendar-check" />
        <span>
          <FormattedMessage id="shared.calendarAvailability" />
        </span>
        <i className="icon icon-right ml-auto" />
      </Button>
      <div
        className={classNames(
          'calendar-listing__side-slideout',
          isOpened && 'calendar-listing__side-slideout--active',
        )}
      >
        <Form layout="vertical">
          <div className="calendar-listing__side-slideout__main">
            <div className="mb-16">
              <BackButton
                isPopoverVisible={isBackPopoverVisible}
                discardPrepare={showPopoverOrBack({
                  popoverName: 'isBackPopoverVisible',
                  props,
                })}
                discardConfirm={closePopoversAndBack({ props })}
                discardReject={closePopover({ popoverName: 'isBackPopoverVisible' })}
              />
            </div>
            <h3 className="text-subheader mb-8">
              <FormattedMessage id="shared.calendarAvailability" />
            </h3>
            <p className="mb-16">
              <FormattedMessage id="calendar.specifyAvailability" />
            </p>
            <Field
              name="countMonth"
              className="main-input__field"
              optionLabelProp="label"
              component={SelectField}
              label={{ id: 'shared.period' }}
              placeholder={{ id: 'shared.select' }}
              items={CALENDAR_AVAILABILITY_OPTIONS}
              formItemClasses="mb-0"
            />
          </div>
          <Footer
            submitBtnText={{ id: 'shared.save' }}
            submit={submitForm}
            isLoading={isLoading}
            canSubmit={isValid}
            isSavePopoverVisible={isSavePopoverVisible}
            isDiscardPopoverVisible={isDiscardPopoverVisible}
            discardPrepare={showPopoverOrBack({
              popoverName: 'isDiscardPopoverVisible',
              props,
            })}
            discardConfirm={closePopoversAndBack({ props })}
            discardReject={closePopover({ popoverName: 'isDiscardPopoverVisible' })}
            withSavePopover
            savePrepare={showPopover({ popoverName: 'isSavePopoverVisible' })}
            saveConfirm={closePopoversAndSave({ props })}
            saveReject={closePopover({ popoverName: 'isSavePopoverVisible' })}
          />
        </Form>
      </div>
    </div>
  );
};

CalendarAvailability.propTypes = {
  toggleOpenedState: PropTypes.func.isRequired,
  submitForm: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  isOpened: PropTypes.bool.isRequired,
  isValid: PropTypes.bool.isRequired,
  isBackPopoverVisible: PropTypes.bool.isRequired,
  isDiscardPopoverVisible: PropTypes.bool.isRequired,
  isSavePopoverVisible: PropTypes.bool.isRequired,
  showPopoverOrBack: PropTypes.func.isRequired,
  closePopover: PropTypes.func.isRequired,
  showPopover: PropTypes.func.isRequired,
  closePopoversAndBack: PropTypes.func.isRequired,
  closePopoversAndSave: PropTypes.func.isRequired,
};

export default CalendarAvailability;
