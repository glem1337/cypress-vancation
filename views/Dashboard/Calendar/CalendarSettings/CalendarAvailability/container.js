import React from 'react';
import PropTypes from 'prop-types';
import { withFormik } from 'formik';
import * as R from 'ramda';
import { connect } from 'react-redux';

import { camperPricingSelector } from 'state/concepts/camper/selectors';
import { CALENDAR_AVAILABILITY_DEFAULT_VALUE } from 'constants/calendar';
import { selectedSlotsSelector } from 'state/concepts/calendar/selectors';
import { handleSubmitWithProps } from 'utils/form/handleSubmit';
import { updateCalendarAvailability } from 'state/concepts/camper/actions';
import { loadingSelector } from 'state/data/selectors';
import { updateCalendarAvailabilityEndpoint } from 'state/concepts/camper/endpoints';
import withPopovers from 'utils/calendar/withPopovers';

import CalendarAvailabilityComponent from './component';

class CalendarAvailability extends React.Component {
  static propTypes = {
    values: PropTypes.shape().isRequired,
    setFieldValue: PropTypes.func.isRequired,
    selectedSlots: PropTypes.shape().isRequired,
    isLoading: PropTypes.bool,
    checkOpenedState: PropTypes.func.isRequired,
    isSubmitting: PropTypes.bool,
  }

  static defaultProps = {
    isLoading: false,
    isSubmitting: false,
  }

  componentDidUpdate(prevProps) {
    const { checkOpenedState } = this.props;

    checkOpenedState(prevProps, this.props);
  }

  render() {
    return (
      <CalendarAvailabilityComponent
        {...this.props}
      />
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  selectedSlots: selectedSlotsSelector(state),
  pricingInfo: camperPricingSelector(state, ownProps?.camperId),
  isLoading: loadingSelector(state, updateCalendarAvailabilityEndpoint.endpoint),
});

const mapDispatchToProps = {
  onSubmit: updateCalendarAvailability,
};

export { CalendarAvailability as CalendarAvailabilityContainer };
export default R.compose(
  withPopovers,
  connect(mapStateToProps, mapDispatchToProps),
  withFormik({
    enableReinitialize: true,
    mapPropsToValues: ({ pricingInfo }) => ({
      countMonth: pricingInfo?.calendarAvailability || CALENDAR_AVAILABILITY_DEFAULT_VALUE.value,
    }),
    handleSubmit: handleSubmitWithProps(['camperId']),
  }),
)(CalendarAvailability);
