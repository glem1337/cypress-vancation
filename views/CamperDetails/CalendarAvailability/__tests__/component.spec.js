import React from 'react';
import { shallow } from 'enzyme';

import CalendarAvailability from '../component';

const mockedHookData = {
  containerRef: {},
  searchParams: { dateRange: null },
};
jest.mock('../hook', () => jest.fn(() => mockedHookData));

const mockedAvailabilityData = {
  formatShortWeekday: jest.fn(),
  tileContent: jest.fn(),
  tileDisabled: jest.fn(),
  onDateRangeChange: jest.fn(),
  onActiveStartDateChange: jest.fn(),
  isCalendarFetching: true,
};
jest.mock('utils/hooks/useCalendarAvailability', () => jest.fn(() => mockedAvailabilityData));

describe('CalendarAvailability component tests', () => {
  const component = shallow(<CalendarAvailability />);

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
