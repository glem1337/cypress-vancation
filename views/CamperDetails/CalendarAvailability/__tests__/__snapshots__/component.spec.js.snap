// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarAvailability component tests matches snapshot 1`] = `
<div
  id="details-availability"
>
  <p
    className="text-headline mb-24"
  >
    <Memo(FormattedMessage)
      id="shared.availability"
    />
  </p>
  <div
    className="campervan-availability__calendar-wrapper campervan-availability__calendar-wrapper--loading"
    id="campervan-availability__calendar-wrapper"
  >
    <Calendar
      calendarType="US"
      formatShortWeekday={[MockFunction]}
      maxDate={+275760-09-13T00:00:00.000Z}
      maxDetail="month"
      minDate={0001-01-01T00:00:00.000Z}
      minDetail="century"
      onActiveStartDateChange={[MockFunction]}
      onChange={[MockFunction]}
      returnValue="start"
      selectRange={true}
      showDoubleView={true}
      showNavigation={true}
      showNeighboringMonth={true}
      tileContent={[MockFunction]}
      tileDisabled={[MockFunction]}
      value={null}
    />
    <div
      className="campervan-availability__calendar-loader-wrapper"
    >
      <Spin
        className="search-page__spin"
        indicator={
          <ForwardRef(LoadingOutlined)
            spin={true}
            style={
              Object {
                "fontSize": 88,
              }
            }
          />
        }
        size="default"
        spinning={true}
        wrapperClassName=""
      />
    </div>
  </div>
</div>
`;
